### PR TITLE
[dagster-airlift] Support dag-level overrides in builder

### PR DESCRIPTION
**Pull request by dpeng817**
## Summary & Motivation
Support dag-level override behavior in the airflow builder.

This iteration explicitly duplicates some implementation pathways across dag / task enrichment.

## How I Tested These Changes
A new test for e2e behavior with sensor.
## Changelog
NOCHANGELOG

### CODE REVIEW

**Comment by dpeng817** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25243?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#25205** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25205?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#25193** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25193?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#25168** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25168?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#25160** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25160?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#25159** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25159?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#25158** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25158?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#25243** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25243?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#25242** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25242?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @dpeng817 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: examples/experimental/dagster-airlift/dagster_airlift/core/airflow_defs_data.py [hunk 1]
```diff
@@ -56,5 +58,15 @@ def asset_keys_per_dag_handle(self) -> Mapping[DagHandle, AbstractSet[AssetKey]]
                     asset_keys_per_handle[dag_handle].add(spec.key)
         return asset_keys_per_handle
 
+    @cached_property
+    def peered_dag_asset_keys_by_handle(self) -> Mapping[DagHandle, AbstractSet[AssetKey]]:
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/core/airflow_defs_data.py:62]: `by_dag_handle` for clarity IMO

File: examples/experimental/dagster-airlift/dagster_airlift/core/sensor/sensor_builder.py [hunk 2]
```diff
@@ -307,37 +304,59 @@ def build_synthetic_asset_materializations(
...
+    context.log.info(f"Found {len(dagster_runs)} dagster runs for {dag_run.run_id}")
 
     context.log.info(
         f"Airlift Sensor: Found dagster run ids: {[run.run_id for run in dagster_runs]}"
         f" for airflow run id {dag_run.run_id} and dag id {dag_run.dag_id}"
     )
+    synthetic_mats = []
+    # Peered dag-level materializations will always be emitted.
+    synthetic_mats.extend(synthetic_mats_for_peered_dag_asset_keys(dag_run, airflow_data))
+    # If there is a dagster run for this dag, we don't need to synthesize materializations for mapped dag assets.
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/core/sensor/sensor_builder.py:321]: The double negative in the comment makes this a little hard to follow.

# No dagster runs means that the computation that materializes the asset was not proxied to Dagster. Therefore we are going to emit synthetic materializations based on the airflow executoion data.

File: examples/experimental/dagster-airlift/dagster_airlift/core/serialization/compute.py [hunk 3]
```diff
@@ -62,10 +74,20 @@ def asset_key_map(self) -> Dict[str, Dict[str, Set[AssetKey]]]:
...
+    @cached_property
+    def asset_keys_by_dag(self) -> Dict[str, Set[AssetKey]]:
+        """Mapping of dag_id to set of asset_keys mapped from that dag."""
+        asset_key_map: Dict[str, Set[AssetKey]] = defaultdict(set)
+        for spec in self.asset_specs:
+            if is_dag_mapped_asset_spec(spec):
+                for dag_handle in dag_handles_for_spec(spec):
+                    asset_key_map[dag_handle.dag_id].add(spec.key)
+        return asset_key_map
+
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/core/serialization/compute.py:86]: this naming is confusing alongside `asset_keys_in_dag_by_id` IMO. `asset_keys_by_dag` should have a more specific name to indicate only contains DAG to asset mappings

Comment (schrockn) [reply examples/experimental/dagster-airlift/dagster_airlift/core/serialization/compute.py:86]: `asset_keys_by_mapped_dag_id`?

File: examples/experimental/dagster-airlift/dagster_airlift/core/utils.py [hunk 4]
```diff
@@ -82,3 +91,13 @@ def dag_handles_for_spec(spec: AssetSpec) -> Set["DagHandle"]:
...
+
+
+def peered_dag_handles_for_spec(spec: AssetSpec) -> Set["DagHandle"]:
+    from dagster_airlift.core.serialization.serialized_data import DagHandle
+
+    check.param_invariant(is_peered_dag_asset_spec(spec), "spec", "Must be mapped spec")
+    dag_handles = []
+    for dag_handle_dict in spec.metadata[PEERED_DAG_MAPPING_METADATA_KEY]:
+        dag_handles.append(DagHandle(dag_id=dag_handle_dict["dag_id"]))
+    return set(dag_handles)
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/core/utils.py:103]: nit: could be set comprehension

File: examples/experimental/dagster-airlift/dagster_airlift/core/serialization/compute.py [hunk 5]
```diff
@@ -26,34 +30,42 @@ class AirliftMetadataMappingInfo:
     asset_specs: List[AssetSpec]
 
     @cached_property
-    def mapped_asset_specs(self) -> List[AssetSpec]:
+    def task_mapped_asset_specs(self) -> List[AssetSpec]:
         return [spec for spec in self.asset_specs if is_task_mapped_asset_spec(spec)]
 
+    @cached_property
+    def dag_mapped_asset_specs(self) -> List[AssetSpec]:
+        return [spec for spec in self.asset_specs if is_dag_mapped_asset_spec(spec)]
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/core/serialization/compute.py:38]: `mapped_dag_asset_specs`?

**Review by schrockn — CHANGES_REQUESTED** [review]
I think we need a round of naming consistency pass here around "mapped task" and "mapped dag"

File: examples/experimental/dagster-airlift/dagster_airlift/core/serialization/compute.py [hunk 6]
```diff
@@ -62,10 +74,20 @@ def asset_key_map(self) -> Dict[str, Dict[str, Set[AssetKey]]]:
                     asset_key_map[task_handle.dag_id][task_handle.task_id].add(spec.key)
         return asset_key_map
 
+    @cached_property
+    def mapped_asset_keys_by_dag_id(self) -> Dict[str, Set[AssetKey]]:
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/core/serialization/compute.py:78]: `asset_keys_by_mapped_dag_id` I think it more accurate?

File: examples/experimental/dagster-airlift/dagster_airlift/core/serialization/compute.py [hunk 7]
```diff
@@ -26,34 +30,42 @@ class AirliftMetadataMappingInfo:
...
+        for dag_id, task_to_asset_map in self.mapped_asset_keys_by_task.items():
+            for asset_keys in task_to_asset_map.values():
+                asset_keys_in_dag_by_id[dag_id].update(asset_keys)
+        for dag_id, asset_keys in self.mapped_asset_keys_by_dag_id.items():
+            asset_keys_in_dag_by_id[dag_id].update(asset_keys)
+        return defaultdict(set, asset_keys_in_dag_by_id)
 
     @cached_property
-    def asset_key_map(self) -> Dict[str, Dict[str, Set[AssetKey]]]:
+    def mapped_asset_keys_by_task(self) -> Dict[str, Dict[str, Set[AssetKey]]]:
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/core/serialization/compute.py:68]: `asset_keys_by_mapped_task_id`?

**Review by schrockn — APPROVED** [review]
Please heed final naming suggestion!
